### DISCUSSION TITLE
What's the best way to close flyout in viewmodel?

### DISCUSSION THREAD

**Discussion by jim-jiang-github** [AvaloniaUI/Avalonia#11466] [Q&A] [answered]
I have a flyout that contains a button. My button is binding to the ViewModel through a Command. I want to close the flyout immediately after clicking the button. What is the most elegant way to achieve this?

**Comment by timunie** [comment, marked as answer]
I think the `View` is allowed to know the `ViewModel`. So what I do in such a case is, that I listen to `Click` in code behind and use the `DataContext` which is my `ViewModel` to execute my command. If that was successful, I can hide the flyout. If that is best practice, I don't know. 

You can also: 
- Use a message bus, e.g.: 
   - https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/mvvm-messenger-and-view-services-in-mvvm
   - https://www.reactiveui.net/docs/handbook/message-bus/
   - https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/messenger
- Use Interaction
   - https://www.reactiveui.net/docs/handbook/interactions/  
   - https://docs.avaloniaui.net/tutorials/music-store-app/opening-a-dialog
- Use a Dialog-Servie, e.g.:
   - https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/

**Reply by jim-jiang-github** [reply]
Thank you very much, you have a point.

**Reply by thevortexcloud** [reply]
The view can know about the view model (it has to, to actually work). The view model can't know about the view. So there is nothing stopping the view from calling a method on the view model or the view model raising an event which the view is subscribed to. The latter approach is the general idea of ReactiveUI's interactions.

If you don't care about the view model state, you can just subscribe to the click event and close no matter what happens. If you do care, the approach I generally use is a ReactiveUI interaction to indicate that the view model has finished doing what ever it was doing. That way the view model can remain in full control at all times without knowing how the view works. It can also close popups/flyouts whenever wants to without being initiated by the view.

You can also just use some sort of command that supports `CanExecute` properly. Which will disable the button whenever the view model indicates the command can't be executed. Which I believe will also stop click events from working?

Anyway, in general you should avoid relying too much on how the views actually work to reduce view model coupling to the UI framework.
